datastore/bolt/api_key.go

```diff
@@ -2,54 +2,249 @@ package bolt
-	return nil
+func (a *apiKeyRepo) CreateAPIKey(ctx context.Context, apiKey *datastore.APIKey) error {
+	return a.createUpdateApplication(apiKey)
 }
 
-func (e *apiKeyRepo) UpdateAPIKey(ctx context.Context, apiKey *datastore.APIKey) error {
-	return nil
+func (a *apiKeyRepo) createUpdateApplication(apiKey *datastore.APIKey) error {
+	return a.db.Update(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(a.bucketName))
+
+		aJson, err := json.Marshal(apiKey)
+		if err != nil {
+			return err
+		}
+
+		err = b.Put([]byte(apiKey.UID), aJson)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func (a *apiKeyRepo) UpdateAPIKey(ctx context.Context, apiKey *datastore.APIKey) error {
+	return a.createUpdateApplication(apiKey)
+}
+
+func (a *apiKeyRepo) FindAPIKeyByID(ctx context.Context, uid string) (*datastore.APIKey, error) {
+	var apiKey datastore.APIKey
+
+	err := a.db.View(func(tx *bbolt.Tx) error {
+		buf := tx.Bucket([]byte(a.bucketName)).Get([]byte(uid))
+
+		if buf == nil {
+			return datastore.ErrAPIKeyNotFound
+		}
+
+		err := json.Unmarshal(buf, &apiKey)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	return &apiKey, err
 }
 
-func (e *apiKeyRepo) FindAPIKeyByID(ctx context.Context, uid string) (*datastore.APIKey, error) {
-	return nil, nil
+func (a *apiKeyRepo) FindAPIKeyByMaskID(ctx context.Context, maskID string) (*datastore.APIKey, error) {
+	var apiKeys []datastore.APIKey = make([]datastore.APIKey, 0, 1)
+	var apiKey datastore.APIKey
+
+	err := a.db.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(a.bucketName))
+		c := b.Cursor()
+
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			err := json.Unmarshal(v, &apiKey)
+			if err != nil {
+				return err
+			}
+
+			if apiKey.MaskID == maskID {
+				apiKeys = append(apiKeys, apiKey)
+				break
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil && len(apiKeys) == 0 {
+		return &apiKey, err
+	}
+
+	if err == nil && len(apiKeys) == 0 {
+		return &apiKey, datastore.ErrAPIKeyNotFound
+	}
+
+	return &apiKeys[0], err
 }
 
-func (e *apiKeyRepo) FindAPIKeyByMaskID(ctx context.Context, maskID string) (*datastore.APIKey, error) {
-	return nil, nil
+func (a *apiKeyRepo) RevokeAPIKeys(ctx context.Context, uids []string) error {
+	err := a.db.Update(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(a.bucketName))
+		for _, uid := range uids {
+			err := b.Delete([]byte(uid))
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	return err
 }
 
-func (e *apiKeyRepo) RevokeAPIKeys(ctx context.Context, uids []string) error {
-	return nil
+func (a *apiKeyRepo) FindAPIKeyByHash(ctx context.Context, hash string) (*datastore.APIKey, error) {
+	var apiKeys []datastore.APIKey = make([]datastore.APIKey, 0, 1)
+	var apiKey datastore.APIKey
+
+	err := a.db.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(a.bucketName))
+		c := b.Cursor()
+
+		for k, v := c.First(); k != nil; k, v = c.Next() {
```

Comment:
This can be replaced with `b.ForEach()`, I think we try to use the provided abstractions unless we really need to do some custom stuff with the foo-loop

Reply:
I think seeking works here, for each will execute for every single pair in the db even after we find what we want.